python/cucim/src/cucim/skimage/measure/_regionprops_gpu_misc_kernels.py

```diff
@@ -0,0 +1,721 @@
+import math
+
+import cupy as cp
+import numpy as np
+
+from cucim.skimage._vendored import ndimage as ndi, pad
+
+from ._regionprops_gpu_basic_kernels import regionprops_bbox_coords
+from ._regionprops_gpu_intensity_kernels import (
+    _get_intensity_img_kernel_dtypes,
+    get_intensity_measure_kernel,
+)
+from ._regionprops_gpu_utils import _find_close_labels, _get_min_integer_dtype
+
+__all__ = [
+    "regionprops_euler",
+    "regionprops_perimeter",
+    "regionprops_perimeter_crofton",
+]
+
+misc_deps = dict()
+misc_deps["perimeter"] = ["slice"]
+misc_deps["perimeter_crofton"] = ["slice"]
+misc_deps["euler_number"] = ["slice"]
+
+
+def _weighted_sum_of_filtered_image(
+    label_image, max_label, image_filtered, coefs, pixels_per_thread=16
+):
+    """Compute weighted sums of pixels for each label.
+
+    1. Apply the coefs LUT to the filtered image to get a coefficient image
+    2. Sum the values in the coefficient image for each labeled region
+
+    This function is used during computation of the Euler characteristic and
+    perimeter properties.
+
+    Parameters
+    ----------
+    label_image : cupy.ndarray
+        Label image.
+    max_label : int
+        Maximum label value.
+    image_filtered : (M, N) ndarray
+        Filtered image (must have integer values that can be used to index into
+        the coefs LUT)
+    coefs : cupy.ndarray
+        Coefficients look-up table (LUT).
+    pixels_per_thread : int, optional
+        Number of pixels per thread.
+
+    Returns
+    -------
+    output : cupy.ndarray
+        Weighted sum of pixels for each label.
+    """
+    coefs_image = coefs[image_filtered]
+
+    # generate kernel for per-label weighted sum
+    coefs_sum_kernel = get_intensity_measure_kernel(
+        coefs_image.dtype,
+        num_channels=1,
+        compute_num_pixels=False,
+        compute_sum=True,
+        compute_sum_sq=False,
+        compute_min=False,
+        compute_max=False,
+        pixels_per_thread=pixels_per_thread,
+    )
+
+    # prepare output array
+    sum_dtype, _, _, _ = _get_intensity_img_kernel_dtypes(coefs_image.dtype)
+    output = cp.zeros((max_label,), dtype=sum_dtype)
+
+    coefs_sum_kernel(
+        label_image,
+        label_image.size,
+        coefs_image,
+        output,
+        size=math.ceil(label_image.size / pixels_per_thread),
+    )
+    return output
+
+
+@cp.memoize(for_each_device=True)
```

Comment:
Could not specifying the cache size here lead to memory issues with many different parameter combinations?

Reply:
In this case the only items are the floating point data type and the GPU device ID so it should be only a handful of combinations. If you memoize over the exact image shape or something then that could be more problematic.